coordinator/src/main/scala/filodb.coordinator/queryplanner/SingleClusterPlanner.scala

```diff
@@ -268,48 +273,80 @@ class SingleClusterPlanner(val dataset: Dataset,
+      case lp: SeriesKeysByFilters         => materializeSeriesKeysByFilters(qContext, lp, forceInProcess)
+      case lp: ApplyMiscellaneousFunction  => materializeApplyMiscellaneousFunction(qContext, lp, forceInProcess)
+      case lp: ApplySortFunction           => materializeApplySortFunction(qContext, lp, forceInProcess)
+      case lp: ScalarVaryingDoublePlan     => materializeScalarPlan(qContext, lp, forceInProcess)
       case lp: ScalarTimeBasedPlan         => materializeScalarTimeBased(qContext, lp)
-      case lp: VectorPlan                  => materializeVectorPlan(qContext, lp)
+      case lp: VectorPlan                  => materializeVectorPlan(qContext, lp, forceInProcess)
       case lp: ScalarFixedDoublePlan       => materializeFixedScalar(qContext, lp)
-      case lp: ApplyAbsentFunction         => materializeAbsentFunction(qContext, lp)
-      case lp: ApplyLimitFunction          => materializeLimitFunction(qContext, lp)
-      case lp: ScalarBinaryOperation       => materializeScalarBinaryOperation(qContext, lp)
-      case lp: SubqueryWithWindowing       => materializeSubqueryWithWindowing(qContext, lp)
-      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp)
-      case lp: LabelCardinality            => materializeLabelCardinality(qContext, lp)
+      case lp: ApplyAbsentFunction         => materializeAbsentFunction(qContext, lp, forceInProcess)
+      case lp: ApplyLimitFunction          => materializeLimitFunction(qContext, lp, forceInProcess)
+      case lp: ScalarBinaryOperation       => materializeScalarBinaryOperation(qContext, lp, forceInProcess)
+      case lp: SubqueryWithWindowing       => materializeSubqueryWithWindowing(qContext, lp, forceInProcess)
+      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp, forceInProcess)
+      case lp: LabelCardinality            => materializeLabelCardinality(qContext, lp, forceInProcess)
       //case _                               => throw new BadQueryException("Invalid logical plan")
     }
   }
   // scalastyle:on cyclomatic.complexity
-
+  // scalastyle:off method.length
   override def materializeBinaryJoin(qContext: QueryContext,
-                                     lp: BinaryJoin): PlanResult = {
-    val lhs = walkLogicalPlanTree(lp.lhs, qContext)
+                                     lp: BinaryJoin,
+                                     forceInProcess: Boolean): PlanResult = {
+    // TODO: could probably give this section its own method and generalize
+    //   to other non-leaf plans (especially since repeat tree traversals are probably
+    //   no big deal w.r.t. performance).
+
+    // occupied with a shard's dispatcher iff the join only spans the single shard
+    var singleShardDispatcherOpt: Option[PlanDispatcher] = None
+
+    if (qContext.plannerParams.targetSchema.isDefined) {
+      val leafPlans = findLeafLogicalPlans(lp.lhs) ++ findLeafLogicalPlans(lp.rhs)
```

Comment:
This would get expensive in a sufficiently-large tree of binary joins. Is it right to assume this will likely never happen?

Reply:
outdated